File: samples/CocosDragonJS/Resources/MultiTouchTest.js
/****************************************************************************
 Copyright (c) 2010-2012 cocos2d-x.org
 Copyright (c) 2008-2010 Ricardo Quesada
 Copyright (c) 2011      Zynga Inc.

 http://www.cocos2d-x.org

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in
 all copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
 ****************************************************************************/

var touchColors = [
    cc.yellow(),
    cc.c3b(0,0,255),
    cc.c3b(0,255,0),
    cc.c3b(255,0,0),
    cc.magenta()
];

var TouchPoint = cc.Node.extend({
    _touchPoint:null,
    _tonchColor:null,
    _styleStr:null,

    ctor:function () {
        this._touchPoint = cc.p(0, 0);
        this._tonchColor = new cc.Color3B(255, 255, 255);
        this._styleStr = "rgba(" + this._tonchColor.r + "," + this._tonchColor.g + "," + this._tonchColor.b + ", 1)";
        //this.setShaderProgram(cc.ShaderCache.getInstance().programForKey(kCCShader_PositionTextureColor));
    },

    draw:function () {
        var context = cc.renderContext;
        context.save();
        context.fillStyle = this._styleStr;
        context.strokeStyle = this._styleStr;
        context.lineWidth = "10";
        var size = this.getContentSize();
        cc.drawingUtil.drawLine(cc.p(0, this._touchPoint.y), cc.p(size.width, this._touchPoint.y));
        cc.drawingUtil.drawLine(cc.p(this._touchPoint.x, 0), cc.p(this._touchPoint.x, size.height));
        cc.drawingUtil.drawPoint(this._touchPoint, 15);
        context.restore();
    },

    setTouchPos:function (pt) {
        this._touchPoint = pt;
    },

    setTouchColor:function (color) {
        this._tonchColor = color;
        this._styleStr = "rgba(" + this._tonchColor.r + "," + this._tonchColor.g + "," + this._tonchColor.b + ", 1)";
    }
});

TouchPoint.touchPointWithParent = function (parent) {
    var ret = new TouchPoint();
    ret.setContentSize(parent.getContentSize());
    ret.setAnchorPoint(cc.p(0, 0));
    return ret;
};

var MultiTouchTestLayer = cc.Layer.extend({
    init:function () {
        if (this._super()) {
            this.setTouchEnabled(true);
            return true;
        }
        return false;
    },

    registerWithTouchDispatcher:function () {
        cc.Director.getInstance().getTouchDispatcher().addStandardDelegate(this, 0);
    },

    _savePointArr:[],
    _isPressed:false,
    onTouchesBegan:function (touches, event) {
        this._isPressed = true;
        if (!touches)
            return;

        this._clearSavePointArr();

        for (var i = 0; i < touches.length; i++) {
            var touch = touches[i];
            var touchPoint = TouchPoint.touchPointWithParent(this);
            var location = touch.getLocation();

            touchPoint.setTouchPos(location);
            touchPoint.setTouchColor(touchColors[i]);

            this.addChild(touchPoint);
            this._savePointArr.push(touchPoint);
        }
    },

    _clearSavePointArr:function () {
        for (var i = 0; i < this._savePointArr.length; i++) {
            this._savePointArr[i].removeFromParentAndCleanup(true);
        }
    },

    onTouchesMoved:function (touches, event) {
        this._clearSavePointArr();
        if (!cc.Browser.isMobile) {
            if (!this._isPressed) {
                return;
            }
        }

        for (var i = 0; i < touches.length; i++) {
            var touch = touches[i];
            var touchPoint = TouchPoint.touchPointWithParent(this);
            var location = touch.getLocation();

            touchPoint.setTouchPos(location);
            touchPoint.setTouchColor(touchColors[i]);

            this.addChild(touchPoint);
            this._savePointArr.push(touchPoint);
        }
    },

    onTouchesEnded:function (touches, event) {
        this._isPressed = false;
        this._clearSavePointArr();
    },

    onTouchesCancelled:function (touches, event) {
        this.onTouchesEnded(touches, event);
    }

});

MultiTouchTestLayer.create = function () {
    var ret = new MultiTouchTestLayer();
    if (ret && ret.init()) {
        return ret;
    }
    return null;
};

var MultiTouchTestScene = TestScene.extend({
    runThisTest:function () {
        var layer = MultiTouchTestLayer.create();
        this.addChild(layer, 0);
        cc.Director.getInstance().replaceScene(this);
    }
});

File: tests/cpp-tests/Classes/MultiTouchTest/MultiTouchTest.cpp
/****************************************************************************
 Copyright (c) 2017-2018 Xiamen Yaji Software Co., Ltd.
 
 http://www.cocos2d-x.org
 
 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:
 
 The above copyright notice and this permission notice shall be included in
 all copies or substantial portions of the Software.
 
 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
 ****************************************************************************/

#include "MultiTouchTest.h"

USING_NS_CC;

MultiTouchTests::MultiTouchTests()
{
    ADD_TEST_CASE(MultiTouchTest);
}

static const Color3B* s_TouchColors[5] = {
    &Color3B::YELLOW,
    &Color3B::BLUE,
    &Color3B::GREEN,
    &Color3B::RED,
    &Color3B::MAGENTA
};

class TouchPoint : public Node
{
public:
    TouchPoint(const Vec2 &touchPoint, const Color3B &touchColor)
    {
        DrawNode* drawNode = DrawNode::create();
        auto s = Director::getInstance()->getWinSize();
        Color4F color(touchColor.r/255.0f, touchColor.g/255.0f, touchColor.b/255.0f, 1.0f);
        drawNode->drawLine(Vec2(0.0f, touchPoint.y), Vec2(s.width, touchPoint.y), color);
        drawNode->drawLine(Vec2(touchPoint.x, 0.0f), Vec2(touchPoint.x, s.height), color);
        drawNode->drawDot(touchPoint, 3, color);
        addChild(drawNode);
    }

    static TouchPoint* touchPointWithParent(Node* pParent, const Vec2 &touchPoint, const Color3B &touchColor)
    {
        auto pRet = new (std::nothrow) TouchPoint(touchPoint, touchColor);
        pRet->setContentSize(pParent->getContentSize());
        pRet->setAnchorPoint(Vec2(0.0f, 0.0f));
        pRet->autorelease();
        return pRet;
    }
};

bool MultiTouchTest::init()
{
    if (TestCase::init())
    {
        auto listener = EventListenerTouchAllAtOnce::create();
        listener->onTouchesBegan = CC_CALLBACK_2(MultiTouchTest::onTouchesBegan, this);
        listener->onTouchesMoved = CC_CALLBACK_2(MultiTouchTest::onTouchesMoved, this);
        listener->onTouchesEnded = CC_CALLBACK_2(MultiTouchTest::onTouchesEnded, this);
        _eventDispatcher->addEventListenerWithSceneGraphPriority(listener, this);
        
        auto title = Label::createWithSystemFont("Please touch the screen!", "", 24);
        title->setPosition(VisibleRect::top()+Vec2(0.0f, -40.0f));
        addChild(title);
        
        return true;
    }
    return false;
}

static Map<int, TouchPoint*> s_map;

void MultiTouchTest::onTouchesBegan(const std::vector<Touch*>& touches, Event  *event)
{
    for ( auto &item: touches )
    {
        auto touch = item;
        auto location = touch->getLocation();
        auto touchPoint = TouchPoint::touchPointWithParent(this, location, *s_TouchColors[touch->getID()%5]);

        addChild(touchPoint);
        s_map.insert(touch->getID(), touchPoint);
    }
}

void MultiTouchTest::onTouchesMoved(const std::vector<Touch*>& touches, Event  *event)
{
    for( auto &item: touches)
    {
        auto touch = item;
        auto pTP = s_map.at(touch->getID());
        auto location = touch->getLocation();
        
        removeChild(pTP, true);
        s_map.erase(touch->getID());
        
        auto touchPointNew = TouchPoint::touchPointWithParent(this, location, *s_TouchColors[touch->getID()%5]);
        addChild(touchPointNew);
        s_map.insert(touch->getID(), touchPointNew);
    }
}

void MultiTouchTest::onTouchesEnded(const std::vector<Touch*>& touches, Event  *event)
{
    for ( auto &item: touches )
    {
        auto touch = item;
        auto pTP = s_map.at(touch->getID());
        removeChild(pTP, true);
        s_map.erase(touch->getID());
    }
}

void MultiTouchTest::onTouchesCancelled(const std::vector<Touch*>& touches, Event  *event)
{
    onTouchesEnded(touches, event);
}
